doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,409 @@
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :return: driver name (e.g. "SPI_0")
+
+.. function:: codegen.device_tree_controller_driver_name_indirect(device_type_name, device_index, property_name)
+
+     Get the driver name of another device given by the property.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param device_index: Index of device
+    :param property_name: Property that denotes the device tree prefix of the other device
+    :return: driver name of other device (e.g. "GPIOA")
+
+Guarding chunks of source code
+------------------------------
+
+.. function:: codegen.if_device_tree_controller_compatible(device_type_name, compatible)
+
+    Stop code generation if there is no activated device that is compatible.
+
+    Code generation stops right before the generator end marker @code{.codeins}@endcode.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param compatible: driver compatibility string (e.g. st,stm32-spi-fifo)
+
+.. function:: codegen.outl_guard_device_tree_controller(device_type_name, compatible)
+
+    Write a guard (#if [guard]) C preprocessor directive to output.
+
+    If there is an activated device that is compatible the guard value is set to 1,
+    otherwise it is set to 0.
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param compatible: driver compatibility string (e.g. st,stm32-spi-fifo)
+
+.. function:: codegen.outl_unguard_device_tree_controller(device_type_name, compatible)
+
+    Write an unguard (#endif) C preprocessor directive to output.
+
+    This is the closing command for codegen.outl_guard_device_tree_controller().
+
+    :param device_type_name: Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    :param compatible: driver compatibility string (e.g. st,stm32-spi-fifo)
+
+.. function:: codegen.outl_guard_config(property_name)
+
+    Write a guard (#if [guard]) C preprocessor directive to output.
+
+    If there is a configuration property of the given name the property value
+    is used as guard value, otherwise it is set to 0.
+
+    :param property_name: Name of the configuration property.
+
+.. function:: codegen.outl_unguard_config(property_name)
+
+    Write an unguard (#endif) C preprocessor directive to output.
+
+    This is the closing command for codegen.outl_guard_config().
+
+    :param property_name: Name of the configuration property.
+
+Logging
+-------
+
+.. function:: codegen.log(message [, message_type=None] [, end="\n"] [, logonly=True])
+
+.. function:: codegen.msg(s)
+
+.. function:: codegen.error(msg='Error raised by codegen.')
+
+    The codegen.error function.
+
+    Instead of raising standard python errors, codegen generators can use
+    this function.  It will display the error without a scary Python
```

Comment:
Again, I'd avoid using "scary Python traceback", so just write this as:

    Instead of raising standard Python errors, codegen generators can use
    this function to write the error message to the log.